alphacsc/_z_encoder.py

```diff
@@ -372,20 +365,25 @@ def get_max_error_patch(self):
 
         n_channels = self.X.shape[1]
         *_, n_times_atom = get_D_shape(self.D_hat, n_channels)
-        return self.X[n0, :, t0:t0 + n_times_atom][None]
+
+        patch = self.X[n0, :, t0:t0 + n_times_atom][None]
+        if self.D_hat.ndim == 2:
+            patch = get_uv(patch)
+        return patch
 
     def set_D(self, D):
         self.D_hat = D
 
-    def set_reg(self, reg):
-        self.reg = reg
+    def update_z_hat(self, D):
+        self.D_hat = D
+        # XXX would it require a more complex check or adjustment. We are
+        # assuming that the number of atoms increase one by one.
+        # check shape and adjust z_hat
 
-    def add_one_atom(self, new_atom):
-        assert new_atom.shape == (self.n_times_atom + self.X.shape[1],)
-        self.D_hat = np.concatenate([self.D_hat, new_atom[None]])
-        self.z_hat = np.concatenate(
-            [self.z_hat, self._get_new_z_hat(1)], axis=1
-        )
+        if self.z_hat.shape[1] < self.n_atoms:
+            self.z_hat = np.concatenate(
+                [self.z_hat, self._get_new_z_hat(1)], axis=1
+            )
```

Comment:
I would use a single function `set_D` that checks that `self.z_hat.shape[1] == self.D_hat.shape[1]` and adds columns if needed.

Reply:
Moved update_z_hat contents to set_D.